dg.go

```diff
@@ -1,315 +1,464 @@
+		}
+	}
+	return nil
+}
+
+func (d *directedGraph[NodeType]) PopReadyNodes() []*node[NodeType] {
+	d.lock.Lock()
+	// Save the map to a local variable to minimize time locked.
+	readyMap := d.readyForProcessing
+	d.readyForProcessing = make(map[string]*node[NodeType])
+	d.lock.Unlock()
+	// Move the map nodes to a slice.
+	result := make([]*node[NodeType], len(readyMap))
+	i := 0
+	for _, node := range readyMap {
+		result[i] = node
+		i += 1
+	}
+	return result
 }
 
 type node[NodeType any] struct {
-    deleted bool
-    id      string
-    item    NodeType
-    dg      *directedGraph[NodeType]
+	deleted                 bool
+	id                      string
+	item                    NodeType
+	ready                   bool
+	status                  ResolutionStatus
+	outstandingDependencies map[string]DependencyType
+	dg                      *directedGraph[NodeType]
+}
+
+func (n *node[NodeType]) ID() string {
+	return n.id
+}
+
+func (n *node[NodeType]) Item() NodeType {
+	return n.item
+}
+
+func (n *node[NodeType]) ResolutionStatus() ResolutionStatus {
+	return n.status
+}
+
+func (n *node[NodeType]) IsReady() bool {
+	return n.ready
+}
+
+func (n *node[NodeType]) ResolveNode(status ResolutionStatus) error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	return n.resolveNode(status)
+}
+
+func (n *node[NodeType]) resolveNode(status ResolutionStatus) error {
+	if n.deleted {
+		return ErrNodeDeleted{n.id}
+	}
+	if n.status != Waiting {
+		return ErrNodeResolutionAlreadySet{n.id, n.status, status}
+	}
+	n.status = status
+	if status == Waiting {
+		return nil // Don't propagate a waiting status.
+	}
+	// Propagate to outbound connections.
+	outboundConnections := n.dg.connectionsFromNode[n.ID()]
+	for outboundConnectionID := range outboundConnections {
+		err := n.dg.nodes[outboundConnectionID].dependencyResolved(n.ID(), status)
+		if err != nil {
+			return err
+		}
+	}
+	return nil
+}
+
+func (n *node[NodeType]) Connect(nodeID string) error {
+	return n.dg.connectNodes(n.id, nodeID, AndDependency)
+}
+
+func (n *node[NodeType]) ConnectDependency(fromNodeID string, dependencyType DependencyType) error {
+	return n.dg.connectNodes(fromNodeID, n.id, dependencyType)
 }
 
-func (d *node[NodeType]) ID() string {
-    return d.id
+func (n *node[NodeType]) DisconnectInbound(fromNodeID string) error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
+	if _, ok := n.dg.nodes[fromNodeID]; !ok {
+		return &ErrNodeNotFound{fromNodeID}
+	}
+	if _, ok := n.dg.connectionsToNode[n.id][fromNodeID]; !ok {
+		return &ErrConnectionDoesNotExist{n.id, fromNodeID}
+	}
+	delete(n.dg.connectionsToNode[n.id], fromNodeID)
+	delete(n.dg.connectionsFromNode[fromNodeID], n.id)
+	return nil
 }
 
-func (d *node[NodeType]) Item() NodeType {
-    return d.item
+func (n *node[NodeType]) DisconnectOutbound(toNodeID string) error {
+	n.dg.lock.Lock()
+	defer n.dg.lock.Unlock()
+	if n.deleted {
+		return &ErrNodeDeleted{n.id}
+	}
```

Comment:
The nodes are public. You can get nodes from the DAG and hold onto them. The functions you can call on the node are supposed lock the DAG lock because the nodes back reference the DAG.
I see a few other sections where the lock wasn't acquired in the getters, so I will add those. Done.